### PR TITLE
[ENH] Adds RED CoMETS classifier 

### PR DESCRIPTION
**Pull request by zy18811**
PR adding [RED CoMETS classifier](https://arxiv.org/abs/2307.13679).

Not sure what algorithm type it falls under? So have put into hybrid for now.

Adds a soft dependency to [imbalanced-learn](https://github.com/scikit-learn-contrib/imbalanced-learn). 

### TODO
- [x] Speed up SAX transforms using parallelisation 
- [X] Fix some minor formatting issues
- [x] Add tests
- [x] Add examples in docstrings + update docs/api_reference
- [x]  get_test_params()

### PR checklist
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc).
- [X] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.
- [x] For new estimators, I've added the estimator to the online documentation.

### CODE REVIEW

**Comment by TonyBagnall** [comment]
great, thanks Luca. The categories are necessarily fluid, but I would probably put it in dictionary based?

**Comment by TonyBagnall** [comment]
also, would you like to join our slack @zy18811 https://join.slack.com/t/aeon-toolkit/shared_invite/zt-2466yn6zm-5fV0EW8JjSVUoe2hzJjRMw

**Comment by TonyBagnall** [comment]
could you add some testing code for this? You can use skip test for the soft dependency,  for example, someting like this before each test function

```python
@pytest.mark.skipif(
    not _check_soft_dependencies("imblearn", severity="none"),
    reason="skip test if required soft dependency esig not available",
)
```

**Review by TonyBagnall — APPROVED** [review]
couple of very minor docstring comments, but otherwise this looks great, thank you for the contribution

File: aeon/classification/dictionary_based/_redcomets.py [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+    variant : {1, 2, 3, 4, 5, 6, 7, 8, 9}
+        RED CoMETS variant to use as per [1]_. Defaults to RED CoMETS-3.
+        Variants 4-9 only support multivariate problems.
+    perc_length : int or float, default=5
+        Percentage of time series length used to determinne number of lenses during
+        pair selection.
+    n_trees : int, default=100
+        Number of trees used by each random forest sub-classifier.
+    random_state : int, RandomState instance or None, default=None
+        If `int`, random_state is the seed used by the random number generator;
```
[the remark this answers is not in the document]
Comment (zy18811) [reply aeon/classification/dictionary_based/_redcomets.py]: Is that for everything quoted in docstrings, e.g. the "RandomState" and "None" in following lines, as well or just _int_?

 NB: copied that phrasing from the HC2 docstring so will need fixing there + anywhere else its used. Is it worth me raising an issue for this?

**Comment by TonyBagnall** [comment]
this test failure is nothing to do with your PR, just a weird tensorflow thing

File: aeon/classification/dictionary_based/_redcomets.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zy18811) [reply aeon/classification/dictionary_based/_redcomets.py]: Okay think I've added double tick quotes everywhere they're appropriate! Raised an [issue](https://github.com/aeon-toolkit/aeon/issues/809) :)

**Comment by zy18811** [comment]
Think I've addressed the two docstring issues :)  🚀

File: aeon/classification/dictionary_based/_redcomets.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+        sfa_clfs :
+            List of ``(RandomForestClassifier(), weight)`` tuples fitted on `SFA`
+            transformed training data
+        sax_transforms :
+            List of ``SAX()`` instances with random word length and alpabet size
+        sax_clfs :
+            List of ``(RandomForestClassifier(), weight)`` tuples fitted on `SAX`
+            transformed training data
+        """
+        from imblearn.over_sampling import SMOTE, RandomOverSampler
```
Comment (TonyBagnall) [inline aeon/classification/dictionary_based/_redcomets.py:204]: with soft dependencies we do a check before the import to give a more informative error. Look in clustering/_k_shapes.py for an example of how its done, but basically import  _check_soft_dependencies at the top then call it like this prior to the import
       _check_soft_dependencies("imblearn", severity="error")

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
one more tiny thing on the soft dependencies and I think its ready, great job

File: aeon/classification/dictionary_based/_redcomets.py (hunk 2 above)
Comment (zy18811) [reply aeon/classification/dictionary_based/_redcomets.py:204]: Yep, currently that check is currently in `__init__()`, should it be moved to be directly before the import statement?

Comment (TonyBagnall) [reply aeon/classification/dictionary_based/_redcomets.py:204]: ah sorry, didnt look closely enough, but yes I think we normally put it in the function with the import.

Comment (zy18811) [reply aeon/classification/dictionary_based/_redcomets.py:204]: Okay moved it to the correct location.

**Review by TonyBagnall — DISMISSED** [review]
LGTM

File: .all-contributorsrc [hunk 3]
```diff
@@ -2189,7 +2189,18 @@
...
+      "login": "zy18811",
+      "name": "Luca Bennett",
+      "avatar_url": "https://avatars.githubusercontent.com/u/74973887?v=4",
+      "profile": "https://github.com/zy18811",
+      "contributions": [
+        "code",
+        "doc",
+        "test"
+      ]
+    },
```
Comment (MatthewMiddlehurst) [inline .all-contributorsrc:2203]:
```suggestion
    }
```
Annoyingly this comma breaks the file

**Review by MatthewMiddlehurst — DISMISSED** [review]
Hi Luca, thanks for the contribution. Great work.

One minor change to the all-contributors file would be good, I can monitor and merge after that is complete unless anyone else wants to review and request changes.

**Review by MatthewMiddlehurst — APPROVED** [review]

**Review by TonyBagnall — APPROVED** [review]

**Pull request merged**